examples/platform/silabs/SiWx917/SiWx917/sl_wifi_if.c

```diff
@@ -399,18 +399,14 @@ static void wfx_rsi_save_ap_info() // translation
     }
     wfx_rsi.dev_state |= WFX_RSI_ST_SCANSTARTED;
     sl_status_t status = SL_STATUS_OK;
-#ifndef EXP_BOARD // TODO: this changes will be reverted back after the SDK team fix the scan API
     sl_wifi_scan_configuration_t wifi_scan_configuration = { 0 };
     wifi_scan_configuration                              = default_wifi_scan_configuration;
-#endif
     sl_wifi_ssid_t ssid_arg;
     ssid_arg.length = strlen(wfx_rsi.sec.ssid);
     memcpy(ssid_arg.value, (int8_t *) &wfx_rsi.sec.ssid[0], ssid_arg.length);
     sl_wifi_set_scan_callback(scan_callback_handler, NULL);
-#ifndef EXP_BOARD
-    // TODO: this changes will be reverted back after the SDK team fix the scan API
+    scan_results_complete = false;
```

Comment:
A binary semaphore should replace this whole mechanism. 
Instead of checking on a global boolean variable and then doing some` osThreadYield();` every time you come back to this context until the boolean is true (or timeout).

Just block on the binary semaphore with a max block time of `WIFI_SCAN_TIMEOUT_TICK`. In your  `scan_callback_handler` give the semaphore (from a ISR context). This is more efficient and you won't be switching context for no reason.

Same logic applies to the bg_scan_callback

Reply:
Hi @jmartinez-silabs we have tried with semaphore we are facing the same issue here as well it's getting stuck if we use the semaphore not going ahead with join. 
this is the branch if you want to look into the changes 
https://github.com/shgutte/connectedhomeip/pull/new/bugfix/917_NCP_power_cycle_with_semaphore
tell me if I'm missing something here